File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        low = 0
        maxlength = 0
        char_map = {}
        for high in range(len(s)):
            char_map[s[high]] = char_map.get(s[high],0)+1
            k = high-low+1
            while len(char_map)<(high-low+1):
                char_map[s[low]] -=1
                if char_map[s[low]]==0:
                    del char_map[s[low]]
                low +=1    
            maxlength = max(maxlength,high-low+1)
        return maxlength    

        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int low = 0;
        int maxlength = 0;
        Map<Character, Integer> map = new HashMap<>();

        for (int high = 0; high < s.length(); high++) {
            char rightChar = s.charAt(high);
            map.put(rightChar, map.getOrDefault(rightChar, 0) + 1);

            // If map.size() < window size, a duplicate character exists inside the window
            while (map.size() < high - low + 1) {
                char leftChar = s.charAt(low);
                map.put(leftChar, map.get(leftChar) - 1);
                if (map.get(leftChar) == 0) {
                    map.remove(leftChar);
                }
                low++; // Move left boundary forward
            }

            // Update maximum valid window size
            maxlength = Math.max(maxlength, high - low + 1);
        }

        return maxlength;
    }
}